### PR TITLE
🐛  Use anchor tags outside tooltip for linking

### LINKED ISSUE #32060: amp-analytics linkers config not working on all web stories links
## What's the issue?

I'm trying to share `CLIENT_ID` from amp cache pages to our primary domain. I believe this is the intended use-case for setting `linkers` in the `amp-analytics` config. We have our own analytics package and not using google analytics.

outbound links in the lower 20% of the `amp-story-page` seem to work correctly and the linker id values set in the config work. any links not in the lower 20% that require a click confirmation, however, do not include the linker id data.

## How do we reproduce the issue?

I have only been able to reproduce this with pages served from the amp cache since, i believe, `linkers` only works there.

1. Visit https://www-jumprope-com.cdn.ampproject.org/c/s/www.jumprope.com/g/most-amazing-street-art-in-san-francisco-/IRYTeNmx
2. On the first page, click on "View my profile" at the bottom and see that the linkers data is added to the url on the destination
3. Go back to the first page and click on the username, "thenextsomewhere" in the upper right (it should be the exact same url as the "view my profile link"), but see that once you click the confirmation and go to the user profile page on jumprope.com, the linkers data was not added.
4. Similar behavior happens on the last card, as well.

## What browsers are affected?

i tried on a few different browsers and saw the issue on all i tried:

* chrome mac os
* safari mac os
* safari ios
* google app ios

## Which AMP version is affected?

`2012301722001`

I just implemented `linkers` today, so i'm not sure how long this has been an issue.

### PR DESCRIPTION
**Pull request by mszylkowski**
Linker cannot access anchor tag from tooltip since it's inside the shadow dom, so we create an anchor outside the shadow root and click it when the tooltip is clicked. Creating the tag and clicking it is considered a trusted user interaction since it's called inside the original click event listener

Demo at https://stories-demos-matias.web.app/examples/amp-story/analytics.html#page=p3. Clicking on the bottom and top links adds the cid to the url

Closes #32060

### CODE REVIEW

File: extensions/amp-story/1.0/amp-story-embedded-component.js [hunk 1]
```diff
@@ -1455,6 +1456,21 @@ export class AmpStoryEmbeddedComponent {
...
+  /**
+   * Linkers don't work on shadow root elements so we need to create an anchor outside it.
+   * @param {!Event} event
+   * @private
+   */
+  onAnchorClick_(event) {
+    event.preventDefault();
+    const outerAnchor = this.win_.document.createElement('a');
+    outerAnchor.href = this.tooltip_.href;
+    outerAnchor.target = this.tooltip_.target;
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/amp-story-embedded-component.js:1468]: There are a bit more parameters we need to support for analytics, can we clone the `a` node using `cloneNode` instead?

Can you check or upload a demo to verify this against many OS/browsers?

Comment (mszylkowski) [reply extensions/amp-story/1.0/amp-story-embedded-component.js:1468]: Sounds good, I'll use `cloneNode` and set up a demo. The programmatic click works on Chrome/Mac but we can test it on other platforms as well

**Comment by newmuis** [comment]
High-level directional question: Is it possible to modify linker to be able to inspect links in shadow roots?  The approach in this PR means we have to manually do this for all links in shadow DOM, and seems prone to us missing one somewhere (which, we're already missing some).

**Comment by mszylkowski** [comment]
From the current way of the linker handles clicks, it seems that you'd need to attach a click listener to all the shadow roots, which is not easy because there's no way to know when they are created (or easily target them with CSS). We'd need extra code in the linker and each shadow dom, so it's more complex in the end. If there are other places where this happens, I could fix those in this PR as well

**Comment by newmuis** [comment]
☹️ ok

Info dialog, bookend (although deprecated), consent dialog, access (maybe?  can't remember), and interactive components, and inline page attachment come to mind

**Comment by mszylkowski** [comment]
Oh yeah, looks like it's a number of places where users could potentially use the linker. However, I'd say most of these cases we would (on purpose) prevent linking for privacy reasons or just lack of usefulness (or being deprecated). Enabling linking on specific scenarios might work better than just enabling them on all (and we also avoid more code complexity for now), and we can start adding this functionality where users see a benefit and where we think it's a good use-case. How does that sound?

**Comment by newmuis** [comment]
Ok, diving deeper into that list, I think we can skip bookend (because it's deprecated) and access (it doesn't actually have a shadow root, which is surprising to me, but... 🤷‍♂️ )

The others, it seems like we should probably do.

Can you factor out the cloning into a util?  Given that it's hacky, I wouldn't be surprised if we find problems with it down the line in some specific device/browser/content combination, and it'd be nice to not have to update every call site.

**Comment by mszylkowski** [comment]
Created a function `clickAnchorOnDom` so you can pass the original anchor and the context you want to use to click it (which will be the parent of the cloned node).

**Comment by Enriqe** [comment]
> Could not test linker working properly since the localhost context doesn't trigger it

I think you should be able to test with the demo page you provided, no? It doesn't work there either?

**Comment by mszylkowski** [comment]
Demo now works perfectly, I had to add the `"proxyOnly": false` to the linker which is not documented on amp.dev (had to find it in the code of `linker-manager.js`) so it triggers on non-cached sites.

File: extensions/amp-story/1.0/utils.js [hunk 2]
```diff
@@ -285,3 +285,17 @@ export function setTextBackgroundColor(element) {
...
   });
 }
+
+/**
+ * Click a clone of the anchor passed in the context of the domElement.
+ * Used to apply linker logic on shadow-dom anchors
+ * @param {!Element} anchorElement
+ * @param {!Element} domElement
+ */
+export function clickAnchorOnDom(anchorElement, domElement) {
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/utils.js:295]: Nit: `triggerClickFromLightDom` or anything more obvious about what this does

**Review by gmajoulet — APPROVED** [review]

**Comment by gmajoulet** [comment]
Enrique and Matias, since we do `preventDefault()` on the actual `<a`, can you double check the eventual amp-analytics implications? Is there anything we could be breaking?

**Comment by mszylkowski** [comment]
Amp-analytics is [triggered at the same time](https://github.com/ampproject/amphtml/pull/32181/files#diff-7fc58bd199f74f6fa70a89239d1592d57c4634e34cf6004315bfff56563594f8R723) we do this, so they don't seem to clash from what I've seen. If other click events were added to the shadow dom that wanted to catch this click, that could be a problem. But on the analytics side it shouldn't break anything since they don't have access to the shadow dom (which is why this bug occurred)

**Comment by mszylkowski** [comment]
I'll add this functionality into `amp-story-consent`, `info-dialog` (from amp-viewer) and `page-attachment` on a followup PR

**Comment by gmajoulet** [comment]
`page-attachment` inline content or link `<a` are not in a Shadow DOM, it should not need any further fix

**Pull request merged**